### PR TITLE
Create a test for block size

### LINKED ISSUE #482: Create large block and transaction test cases
As part of block and transaction validation (#428), we'd like to test large blocks and transactions. 

Since Sapling, the maximum block and transaction size is 2,000,000 bytes (MAX).

Note that the maximum size of a transaction is actually slightly lower, because the block also contains a block header, and a transaction count `compact size` field. This effective limit isn't documented in the spec.

The same restriction doesn't apply to mempool transactions, because protocol messages can be up to 2 MB (2 * 1024 * 1024). This inconsistency between mempool and block transactions isn't documented in the spec.

We would like the following test cases:

Block parsing:
* a MAX size block
* a (MAX+1) size block

Mempool transaction parsing:
* ~a MAX size transaction~ (moved to #571)
* ~a (MAX+1) size transaction~ (moved to #571)
* a transaction that's the largest size that will fit in a MAX size block (moved mempool testing to #571)
* a transaction that's 1 byte larger than the largest size that will fit in a MAX size block (moved mempool testing to #571)

These blocks and transactions should use the latest block and transaction formats (Blossom or Heartwood).

Our property tests may already cover these cases, but they are edge cases, so we'd like to be sure.

This issue is open to external contributors. But the Zebra team might also create some of these test cases ourselves.

### PR DESCRIPTION
**Pull request by oxarbitrage**
Attempt to do https://github.com/ZcashFoundation/zebra/issues/482

### CODE REVIEW

File: zebra-chain/src/block/tests.rs [hunk 1]
```diff
@@ -130,6 +133,137 @@ fn deserialize_block() {
...
+    // 23511 transactions = data.len() = 1999925 bytes
+    // 23512 transactions = data.len() = 2000010 bytes
+    let mut limit_txs = 23511;
+
+    // Create 23511 transactions, this is the limit with our DUMMY_TX1 data
+    let mut many_transactions = Vec::new();
+    for _n in 1..limit_txs+1 {
+        let trx = Transaction::zcash_deserialize(&zebra_test::vectors::DUMMY_TX1[..]);
+        many_transactions.push(Arc::new(trx.unwrap()));
+    }
```
Comment (yaahc) [inline zebra-chain/src/block/tests.rs]: TIOLI: you might be able to simplify this a little with `collect` and `repeat`

```rust
let trx = Transaction::zcash_deserialize(&zebra_test::vectors::DUMMY_TX1[..]).unwrap();
let many_transactions = std::iter::repeat(trx.clone())
    .take(limit_txs)
    .collect::<Vec<_>>();
```

File: zebra-chain/src/block/tests.rs [hunk 2]
```diff
@@ -130,6 +133,137 @@ fn deserialize_block() {
...
+
+    // Deserialize by now is ok as we are lower than the limit
+    let block2 = Block::zcash_deserialize(&data[..]).expect("block should deserialize as we are just below limit");
+
+    assert_eq!(block, block2);
+    assert_eq!(limit_txs, block2.transactions.len());
+
+    // Add 1 more transaction to the block, limit will be reached
+    limit_txs = limit_txs + 1;
+    let trx = Transaction::zcash_deserialize(&zebra_test::vectors::DUMMY_TX1[..]);
```
Comment (yaahc) [inline zebra-chain/src/block/tests.rs]: assuming `Transaction` is clone you might want to create a single trx at the top of the file and then clone it everywhere you need it.

File: zebra-chain/src/block/tests.rs [hunk 3]
```diff
@@ -130,6 +133,137 @@ fn deserialize_block() {
...
+    // Create a big transaction
+    let big_transaction = Transaction::V1 {
+        inputs: inputs.clone(),
+        outputs: outputs.clone(),
+        lock_time: LockTime::Time(DateTime::<Utc>::from_utc(NaiveDateTime::from_timestamp(61, 0), Utc)),
+    };
+
+    // Put the big transaction into a block
+    let mut transactions = Vec::new();
+    transactions.push(Arc::new(big_transaction.clone()));
```
Comment (yaahc) [inline zebra-chain/src/block/tests.rs]:
```suggestion
    let transactions = vec![Arc::new(big_transaction.clone())];
```

**Review by yaahc — DISMISSED** [review]
This looks great, I left some comments for some things I think you could refactor to be closer to "idiomatic rust" but none of those are critical issues, the core functionality these tests are exercising looks good to me.

File: zebra-chain/src/block/tests.rs [hunk 4]
```diff
@@ -130,6 +133,137 @@ fn deserialize_block() {
         .expect("block test vector should deserialize");
 }
 
+#[test]
+fn block_limits_multi_tx() { // Test multiple small transactions to fill a block max size
+    let blockheader = generate_block_header();
+
+    // 23511 transactions = data.len() = 1999925 bytes
+    // 23512 transactions = data.len() = 2000010 bytes
+    let mut limit_txs = 23511;
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Suggestion: Can you calculate this number based on the maximum block size `MAX_BLOCK_BYTES` ?

If you use @yaahc's suggestion of collect and repeat, you can use a constant:
```suggestion
    let MAX_TRANSACTIONS_IN_BLOCK = MAX_BLOCK_BYTES / zebra_test::vectors::DUMMY_TX1.len();
```

Comment (teor2345) [reply zebra-chain/src/block/tests.rs]: Oh, but you would also need to use the header size.

Maybe don't worry about it, if the calculation is too complex.

File: zebra-chain/src/block/tests.rs [hunk 5]
```diff
@@ -130,6 +133,137 @@ fn deserialize_block() {
...
+    let mut limit_inputs = 48743;
+
+    let mut inputs = Vec::new();
+    let mut outputs = Vec::new();
+
+    // Create 48743 inputs, this is the limit with our DUMMY_INPUT1 data
+    for _n in 1..limit_inputs+1 {
+        let input_s = TransparentInput::zcash_deserialize(&zebra_test::vectors::DUMMY_INPUT1[..]);
+        inputs.push(input_s.unwrap());
+    }
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: You could apply the same changes that @yaahc suggested here as well:
https://github.com/ZcashFoundation/zebra/pull/558/files#r448012897
https://github.com/ZcashFoundation/zebra/pull/558/files#r448013637

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for covering the "multiple transactions" and "large transaction" cases.

Suggestion:
Before we merge, I'd like us to move the large test vector generation to the zebra-tests crate. That way, we can run other tests on large blocks. In particular, we might want to run the large blocks through our validators.

But the block test functions `block_limits_multi_tx()` and `block_limits_single_tx()` should stay in zebra-chain.

Ticket Update:
This PR covers most of #482. I have split the remaining work out into a new ticket #571, because it depends on wallet support.

**Comment by teor2345** [comment]
Suggestion: If you configure your editor to run rustfmt automatically, your code will pass the rustfmt CI test.

File: zebra-chain/src/block/tests.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebra-chain/src/block/tests.rs]: Thanks, applied.

File: zebra-chain/src/block/tests.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-chain/src/block/tests.rs]: I didn't moved it to the start of the file nor made them `const` but moved to the top of the 2 tests and cloned where needed as suggested.

File: zebra-chain/src/block/tests.rs (hunk 3 above)
Comment (oxarbitrage) [reply zebra-chain/src/block/tests.rs]: Thanks! applied.

File: zebra-chain/src/block/tests.rs (hunk 5 above)
Comment (oxarbitrage) [reply zebra-chain/src/block/tests.rs]: Applied the changes to the second test(`block_limits_single_tx`) too. Thanks.

File: zebra-chain/src/block/tests.rs (hunk 4 above)
Comment (oxarbitrage) [reply zebra-chain/src/block/tests.rs]: Applied at https://github.com/ZcashFoundation/zebra/pull/558/commits/58c11c33affafbfecd89e49c940ee8dc3082d4cb

File: zebra-chain/src/block/tests.rs [hunk 6]
```diff
@@ -130,6 +133,167 @@ fn deserialize_block() {
...
+        .expect("Block header should serialize");
+
+    // Calculate the number of transactions we need
+    let mut max_transactions_in_block =
+        (MAX_BLOCK_BYTES as usize - data_header.len()) / *&zebra_test::vectors::DUMMY_TX1[..].len();
+
+    // Create transactions to be just below the limit
+    let mut many_transactions = std::iter::repeat(Arc::new(tx.clone()))
+        .take(max_transactions_in_block)
+        .collect::<Vec<Arc<_>>>();
```
Comment (yaahc) [inline zebra-chain/src/block/tests.rs]: fwiw the `Arc<_>` here can probably be inferred by the compiler
```suggestion
        .collect::<Vec<_>>();
```

File: zebra-chain/src/block/tests.rs [hunk 7]
```diff
@@ -130,6 +133,167 @@ fn deserialize_block() {
...
+
+    assert_eq!(max_transactions_in_block, many_transactions.len());
```
Comment (yaahc) [inline zebra-chain/src/block/tests.rs]: I don't think this can ever fail based on the above construction

File: zebra-chain/src/block/tests.rs [hunk 8]
```diff
@@ -130,6 +133,167 @@ fn deserialize_block() {
...
+        .expect("block should serialize as we are not limiting generation yet");
+
+    // Deserialize will now fail
+    Block::zcash_deserialize(&data[..]).expect_err("block should not deserialize");
+}
+
+#[test]
+fn block_limits_single_tx() {
+    // Test block limit with a big single transaction
+    use crate::transaction::TransparentInput;
```
Comment (yaahc) [inline zebra-chain/src/block/tests.rs]: TIOLI: I'd probably move these to the top of the file

**Review by yaahc — DISMISSED** [review]
This looks great btw, I'm happy to merge this whenever, I left some comments on things but they're mostly informational so if you would rather merge as is lmk and I'll go ahead and close the PR.

**Review by yaahc — DISMISSED** [review]

File: zebra-chain/src/lib.rs [hunk 9]
```diff
@@ -7,14 +7,14 @@
...
 
-mod merkle_tree;
 mod serde_helpers;
 mod sha256d_writer;
 
 pub mod addresses;
 pub mod block;
 pub mod equihash_solution;
 pub mod keys;
+pub mod merkle_tree;
```
Comment (yaahc) [inline zebra-chain/src/lib.rs]: I'm not sure but we probably don't want to make this pub. cc @ZcashFoundation/zebra-team

If this isn't reachable by the `tests::generate` module try making it `pub(crate)`

File: zebra-chain/src/block/tests.rs [hunk 10]
```diff
@@ -165,31 +171,25 @@ fn block_limits_multi_tx() {
         .zcash_serialize(&mut data)
         .expect("block should serialize as we are not limiting generation yet");
 
-    // Check size
-    assert_eq!(data.len(), 1999925);
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Issue: The block size is the purpose of the test, so let's make sure `data.len() <= MAX_BLOCK_BYTES`.

File: zebra-chain/src/block/tests.rs [hunk 11]
```diff
@@ -165,31 +171,25 @@ fn block_limits_multi_tx() {
...
+    assert_eq!(max_transactions_in_block, many_transactions.len());
 
     // Serialize will still be fine
     let mut data = Vec::new();
     block
         .zcash_serialize(&mut data)
         .expect("block should serialize as we are not limiting generation yet");
 
-    // Check size
-    assert_eq!(data.len(), 2000010);
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Issue: The block size is the purpose of the test, so let's make sure `data.len() > MAX_BLOCK_BYTES`.

File: zebra-chain/src/block/tests.rs [hunk 12]
```diff
@@ -248,24 +264,18 @@ fn block_limits_single_tx() {
         .zcash_serialize(&mut data)
         .expect("block should serialize as we are not limiting generation yet");
 
-    // Check size
-    assert_eq!(data.len(), 1999972);
-
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Issue: The block size is the purpose of the test, so let's make sure `data.len() <= MAX_BLOCK_BYTES`.

File: zebra-chain/src/block/tests.rs [hunk 13]
```diff
@@ -280,9 +290,6 @@ fn block_limits_single_tx() {
         .zcash_serialize(&mut data)
         .expect("block should serialize as we are not limiting generation yet");
 
-    // Check size
-    assert_eq!(data.len(), 2000013);
-
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Issue: The block size is the purpose of the test, so let's make sure `data.len() > MAX_BLOCK_BYTES`.

File: zebra-chain/src/block/tests.rs [hunk 14]
```diff
@@ -130,6 +136,160 @@ fn deserialize_block() {
...
+    // Create transactions to be just below the limit
+    let mut many_transactions = std::iter::repeat(Arc::new(tx.clone()))
+        .take(max_transactions_in_block)
+        .collect::<Vec<_>>();
+
+    // Add the transactions into a block
+    let mut block = Block {
+        header: blockheader,
+        transactions: many_transactions.clone(),
+    };
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Suggestion: Can you move large block generation to its own function in a `zebra_chain::test::generate` module?

I'd like to be able to create large blocks in other tests.

The functions would look something like:
```
fn large_multi_transaction_block() -> Block
fn large_single_transaction_block() -> Block
fn oversized_multi_transaction_block() -> Block
fn oversized_single_transaction_block() -> Block
```

There might be some way to share code between these functions, but let's do that in a future PR.

File: zebra-chain/src/block/tests.rs [hunk 15]
```diff
@@ -130,6 +136,160 @@ fn deserialize_block() {
...
+        .expect("block should deserialize as we are just below limit");
+
+    assert_eq!(block, block2);
+    assert_eq!(max_transactions_in_block, block2.transactions.len());
+
+    // Add 1 more transaction to the block, limit will be reached
+    max_transactions_in_block = max_transactions_in_block + 1;
+    many_transactions.push(Arc::new(tx));
+    block.transactions = many_transactions.clone();
+    assert_eq!(max_transactions_in_block, many_transactions.len());
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Suggestion: Can you move this oversize block generation code to `oversize_multi_transaction_block()`  in a `zebra_chain::test::generate` module?

File: zebra-chain/src/block/tests.rs [hunk 16]
```diff
@@ -130,6 +136,160 @@ fn deserialize_block() {
...
+        outputs: outputs.clone(),
+        lock_time: locktime,
+    };
+
+    // Put the big transaction into a block
+    let transactions = vec![Arc::new(big_transaction.clone())];
+    let block = Block {
+        header: blockheader,
+        transactions: transactions.clone(),
+    };
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Suggestion: Can you move this large block generation code to `large_single_transaction_block()`  in a `zebra_chain::test::generate` module?

File: zebra-chain/src/block/tests.rs [hunk 17]
```diff
@@ -130,6 +136,160 @@ fn deserialize_block() {
...
+        outputs: outputs.clone(),
+        lock_time: locktime,
+    };
+
+    let mut transactions = Vec::new();
+    transactions.push(Arc::new(new_big_transaction.clone()));
+    let new_block = Block {
+        header: blockheader,
+        transactions: transactions,
+    };
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs]: Suggestion: Can you move this large block generation code to `oversize_single_transaction_block()`  in a `zebra_chain::test::generate` module?

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for making improvements to the code, I think we're almost there.

Issue: Let's keep the block size asserts, because we need large block sizes for the tests to be useful.

Suggestion: Let's put the test vector generation module in `zebra-chain`, behind `#[cfg(test)]`. It would be nice to have a module path like `zebra_chain::test::generate`.

Suggestion: Let's split out the functions that generate large blocks, so we can re-use them in other tests.

**Comment by oxarbitrage** [comment]
I think i addressed all the issues and suggestions, please re-review when you can.

Little performance impact note: In previous version we were adding transactions(or inputs) up to just below the limit and then adding 1 more to "overflow". In this new version we go below the limit with 1 function then go beyond by filling the whole thing again.

File: zebra-chain/src/test.rs [hunk 18]
```diff
@@ -0,0 +1,3 @@
+//! Something
```
Comment (yaahc) [inline zebra-chain/src/test.rs]:
```suggestion
//! Test support code
```

File: zebra-chain/src/lib.rs (hunk 9 above)
Comment (oxarbitrage) [reply zebra-chain/src/lib.rs]: Thanks, this is actually not needed in current version. Reversed at at https://github.com/ZcashFoundation/zebra/pull/558/commits/3df9c5de6b3d7fc4cab6afa40261b589749d8f74

File: zebra-chain/src/test.rs (hunk 18 above)
Comment (oxarbitrage) [reply zebra-chain/src/test.rs]: Thanks, added at https://github.com/ZcashFoundation/zebra/pull/558/commits/3df9c5de6b3d7fc4cab6afa40261b589749d8f74

File: zebra-chain/src/test/vectors.rs [hunk 19]
```diff
@@ -0,0 +1,10 @@
+//! Test vectors for blockchain contructions
+use hex::FromHex;
+
+use lazy_static::lazy_static;
+
+lazy_static! {
```
Comment (teor2345) [inline zebra-chain/src/test/vectors.rs]: Question: these test vectors don't depend on zebra-chain. Would they be more useful in zebra-test?

File: zebra-chain/src/block/tests.rs [hunk 20]
```diff
@@ -130,6 +121,74 @@ fn deserialize_block() {
...
+    // Deserialize will now fail
+    Block::zcash_deserialize(&data[..]).expect_err("block should not deserialize");
+}
+
+#[test]
+fn block_limits_single_tx() {
+    // Test block limit with a big single transaction
+
+    // Create a block just below the limit
+    let mut block = generate::large_single_transaction_block();
```
Comment (teor2345) [inline zebra-chain/src/block/tests.rs:164]: This is great! It's very easy to read the tests now, and see what they are going to test.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks, looks like we're almost done here.

I just have a question about which module the static test vectors should be in.

**Comment by oxarbitrage** [comment]
Sorry, i made a few mistakes when fixing conflicts so had to force push. Apart from that i moved all the test vectors to `zebra-test`. I think is better as we are moving everything of that kind there now.

**Review by yaahc — APPROVED** [review]
LGTM :tada:

**Review by teor2345 — APPROVED** [review]
Thanks, looks great!

**Pull request merged**